### ISSUE TITLE
Add first class support for SVG imports and exports

### ISSUE THREAD

**Issue by lostPixels** [processing/p5.js#4630]
#### Most appropriate sub-area of p5.js?

- [ ] Color
- [x ] Core/Environment/Rendering
- [ ] Data
- [ ] Dom
- [ ] Events
- [ ] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] Other (specify if possible)

#### New feature details:

As an artist using p5.js for plotter drawings, I use vector graphics to power my Axidraw. There are very few options for this workflow out there, and p5.js has by far the best API. 

Although I am still able to partially use it, the p5js.svg library was abandoned over 5 years ago. It is the only way to draw and export SVGs. It's plagued with many bugs that require a deep understanding to fix.

It would be amazing if p5.js natively supported SVG operations. The 2D API maps perfectly to it, and the only new option that would be nice is loading and manipulating external SVGs.

**Comment by brunocastro** [comment]
I'm so happy to read this. I will need export to svg on my project.

**Comment by brunocastro** [comment]
I will need to export to svg in my project too. Sometimes, we needed to edit the programmatic art in vector software.

**Comment by jimmymorris** [comment]
+1 for this

**Comment by bobcgausa** [comment]
https://editor.p5js.org/Bobcook47/sketches/Dk3Osbx3z
Can’t help with export
But am lookin for testers on import
Library handles some files and fragments

**Comment by klavsbuss** [comment]
+1

**Comment by bobcgausa** [comment]
Try my p5 import library for svg, no export though

https://editor.p5js.org/Bobcook47/present/Dk3Osbx3z

Sent from my iPad

On Feb 20, 2021, at 5:28 PM, Klavs <notifications@github.com> wrote:

﻿

+1

**Comment by nataliefreed** [comment]
+1 on this. Is there a place where this (support for vector exports from P5 for use with pen plotters and other digital fabrication workflows) is being discussed? I know there was a discussion [here](https://github.com/processing/p5.js/issues/458) a while back, and that it's not an easy thing to implement. But not sure if it's a design question or an issue of finding someone with the expertise and time. The workflow for this is so nice in Processing, it would be incredible to have that capability in shareable web apps. There are other Javascript libraries that handle SVGs of course, but there's nothing like P5.js for tinkering and teaching.

**Comment by limzykenneth** [comment]
@nataliefreed No current discussions that I know of is going on though there had been pretty consistent interest in this. I think the lack of progress is mostly due to lack expertise and time. It's not a high priority project for p5.js officially at this time and I don't know of any addon projects that are working on this at the moment.

**Comment by zenozeng** [comment]
@nataliefreed 

Hi, I just released p5.js-svg@1.0.5, which fixes a lot of bugs and should work with p5.js@1.3.1.
- https://github.com/zenozeng/p5.js-svg

And the following are the examples for SVG imports and exports
- http://zenozeng.github.io/p5.js-svg/examples/#exporting
- http://zenozeng.github.io/p5.js-svg/examples/#manipulating

**Comment by jimmymorris** [comment]
@zenozeng thanks for updating! this is super exciting, look forward to trying it out!

**Comment by aafiyahtech** [comment]
Hi @zenozeng Cool, let me check

There could be huge scope if we could import an SVG as a shape and then manipulate it in p5js, as the native shape doesn't properly support the translate function. So moving the shape becomes difficult

**Comment by agrshch** [comment]
It's great to see that the SVG support is added as a proposal for 2.0! 
SVG runtime by @zenozeng is truly amazing, but it doesn't cover all the use cases and it can be significantly slower than regular canvas. Its solvable issues though. I used tricks like drawing everything on the canvas for preview and creating (and deleting) the second p5 instance to export the result as SVG. But it would be really nice to have an opportunity to load and export SVG files without switching runtime. Like in Processing

**Comment by thrly** [comment]
+1 on an officially supported SVG export sometime please 🙏

**Comment by goldwasser** [comment]
Also adding +1 on official support for SVG import/export.

I believe a more fundamental change that might be a precursor to this is support for shapes as first-class objects, somewhat akin to the PShape class in the Java Processing library.  Having shapes as first-class objects would not only be useful/necessary when modeling data associated with an SVG file but would have impact even when not using SVG files.  While createShape/endShape can currently be used to render a custom shape, they do not create a first-class object that can be stored or later manipulated or restyled.

Given the advantages of javascript, it would also be great to allow a PShape-like class register callbacks for mouse events such as onEnter, onExit, onClick, and so those would already provide very useful new functionality for a variety of shapes, helping automate tasks such as determining which "shape" is the topmost one underneath a mouse location.

I don't see any open issues about first-class shapes, and I could split this and start a new issue. But given that this particular issue has been tagged for p5.js 2.0, I thought I'd start by adding to this thread.

**Comment by AaratiAkkapeddi** [comment]
Adding another +1!

**Comment by golanlevin** [comment]
Just in case anyone is searching for an SVG export solution in mid-2025, I wanted to make sure people were aware of [p5.plotSvg](https://github.com/golanlevin/p5.plotSvg), an SVG exporter for p5.js that is specifically intended for pen-plotting. It operates similarly to SVG export in Processing (Java) and (unlike Zenozeng's p5.js-svg) maintains the same runtime display as regular p5.

**Comment by NalinDalal** [comment]
@ksen0 @limzykenneth @davepagurek , is this issue up for grabs, trying to understand the issue from my side and exploring the solutions,
just for clearing my confusion, is the issue talking about something like:
**We want to export a p5.js sketch as an svg and let p5.js to upload a svg and edit that svg based on coding/library just as we do normal sketches??**

**Comment by NalinDalal** [comment]
I'd like to start with SVG export using a snapshot approach: To add a saveSVG() function that captures one frame's draw calls as SVG, using the existing visitor pattern (writing a PrimitiveToSVGConverter alongside [PrimitiveToPath2DConverter](https://github.com/processing/p5.js/blob/f4ce8a006be02949fd8d424ed602fefc3ffe42a2/src/shape/custom_shapes.js#L1034)). This would work alongside the normal canvas renderer without replacing it. Would this approach align with what maintainers have in mind?

**Comment by NalinDalal** [comment]
Hi all, following maintainer guidance, I'm starting work on the SVG export addon using the p5.js-addon-template. The plan is:

- Build shape export using the visitor pattern (beginShape/endShape).
- Research how to include text export (noting possible complications).
- Decide if renderer interception is needed for full coverage.
- Keep all technical discussions and progress updates here for visibility.

If anyone has suggestions or requirements for text export, or other features, please comment below!

[Forked Repo for SVG Export](https://github.com/NalinDalal/p5.js-svg-export)

**Comment by NalinDalal** [comment]
My current plan for the add-on:
- SVG export via a `saveSVG()` snapshot approach (capturing one frame's draw calls using the visitor pattern, similar to `PrimitiveToPath2DConverter`)
- SVG import: parsing SVG elements into p5.js-compatible shapes
- Examples showing round-trip usage (import → manipulate → export)

I'll use it hands-on with some plotter-style sketches to gather observations about what works well and what gaps remain, and will report back here.

**Comment by golanlevin** [comment]
Hi @NalinDalal, 

Kindly check out the following SVG exporter for p5.js before you get too far: 
* p5.plotSvg, https://github.com/golanlevin/p5.plotSvg

I've been working on this for some time; it currently works for p5.js 1.1x and I am nearly done updating it for p5.js v2.0. It operates the way you describe, using a snapshot approach rather than replacing the renderer. I modeled it after beginRecord/endRecord from Processing.

It has some limitations at the moment; it does not handle SVG export of pixel-based images, transparency, filters, shaders, blend modes, gradients, animation, or fills and stroke weights. Some of these would be easy to add (fills, stroke weights), but others would be more complex (images, shaders). 

Perhaps it makes sense for us to combine efforts? I could use some help: 
* Handling text (using text-to-points or something similar, as described [here](https://github.com/golanlevin/p5.plotSvg/issues/15)
* Converting it into a proper p5.js addon

Best
@golanlevin

**Comment by NalinDalal** [comment]
Hi @golanlevin thanks for the heads up, literally saves our time
Let me properly check into that, how do you suggest we move forward, do I fork your repo?
Also am a bit confused on the maintenance of the library as mentioned by kit in the discord

<img width="1025" height="368" alt="Image" src="https://github.com/user-attachments/assets/83a0b05b-726f-4773-8694-a82d74b59a78" />

**Comment by golanlevin** [comment]
Hi @NalinDalal, 
Thanks for your reply. I'm giving [p5.plotSvg](https://github.com/golanlevin/p5.plotSvg/) a thorough update tonight to see if it can finally have good compatibility with both v1 and v2 versions of p5.js. I will also turn it into a proper add-on, which is overdue.

I believe there are some major issues to discuss and decide about SVG exporting for p5.js: 
* What's the _goal_ — is it (**A**) to specifically support pen-plotters, which only require a limited subset of p5 graphics, or (**B**) to accurately reproduce the graphics produced by p5, but in (mostly) vector format? If the goal is (A), then p5.plotSvg is a good starting point; if it is (B), then @zenozeng's [p5.js-svg](https://github.com/zenozeng/p5.js-svg) could have more answers, since they have put real effort into making the results 'pixel-perfect'.
* Should SVG exporting ultimately be part of p5.js itself, or reside in an add-on? 

One of the main complications is that SVG and HTML Canvas graphics are _superficially_ similar, but actually quite different in some important details. A good example is discussed [here](https://github.com/golanlevin/p5.plotSvg/issues/27), related to p5's [`clip()`](https://p5js.org/reference/p5/clip/) function. Implementing an SVG exporter that can reproduce what `clip()` does would be extremely challenging, requiring a ton of computational geometry.

I think an acceptable middle ground could be to make an SVG exporter that supports more than just pen-paths, but still has "known limitations" that throw warnings to the user. Such a library (likely an addon) would be unable to reproduce `clip()` in the near term, for example, but would handle visual stylings like fills, strokeWeights, and transparency, as @mariuswatz requested [here](https://github.com/golanlevin/p5.plotSvg/issues/7).

More soon,
Best
Golan

**Comment by bobcgausa** [comment]
Take a look at my SVG import library.  Just search for SVG.  There are many many test cases.
https://editor.p5js.org/Bobcook47/sketches
Maybe it will save you some time
________________________________

**Comment by limzykenneth** [comment]
To clarify the current plan for SVG in p5.js (@ksen0 @davepagurek feel free to correct if I'm wrong), we are happy to see efforts in exploring SVG support in p5.js within addons, I believe addons are the tool for us to experiment with ideas without the pressure of buy-in with a potential breaking change down the line. That being said, it is also encouraged to look at existing efforts in the community for these kind of ideas and only create new addons if there are new ideas especially around API design that could be incompatible.

Internally, the plan for SVG within p5.js core is something I have thought about for a long time and part of the changes made in v2 is towards that (or at least to unblock some of it). I am personally keeping an eye on p5.js-svg by @zenozeng which has good promising implementation for an SVG mode. That being said, we currently have other priorities we'd like to work on such as WebGPU mode, p5.strands, and an in the works FES update, and don't have the capacity to scope out SVG in full. I'd want to fully implement SVG in p5.js as much as I want to do it justice. Having contributors like everyone here exploring the possibilities addons will certainly speed up the process once we got around to it! We may not explicitly review any particular implementation directly, but at least personally I often go to different projects quietly just to check things out for ideas, I may not leave any comments but I always learn a lot from many other projects.

@golanlevin Let me know if you need help with p5.plotSvg for v2!

**Comment by NalinDalal** [comment]
> Hi [@NalinDalal](https://github.com/NalinDalal), Thanks for your reply. I'm giving [p5.plotSvg](https://github.com/golanlevin/p5.plotSvg/) a thorough update tonight to see if it can finally have good compatibility with both v1 and v2 versions of p5.js. I will also turn it into a proper add-on, which is overdue.
> 
> I believe there are some major issues to discuss and decide about SVG exporting for p5.js:
> 
> * What's the _goal_ — is it (**A**) to specifically support pen-plotters, which only require a limited subset of p5 graphics, or (**B**) to accurately reproduce the graphics produced by p5, but in (mostly) vector format? If the goal is (A), then p5.plotSvg is a good starting point; if it is (B), then [@zenozeng](https://github.com/zenozeng)'s [p5.js-svg](https://github.com/zenozeng/p5.js-svg) could have more answers, since they have put real effort into making the results 'pixel-perfect'.
> * Should SVG exporting ultimately be part of p5.js itself, or reside in an add-on?
> 
> One of the main complications is that SVG and HTML Canvas graphics are _superficially_ similar, but actually quite different in some important details. A good example is discussed [here](https://github.com/golanlevin/p5.plotSvg/issues/27), related to p5's [`clip()`](https://p5js.org/reference/p5/clip/) function. Implementing an SVG exporter that can reproduce what `clip()` does would be extremely challenging, requiring a ton of computational geometry.
> 
> I think an acceptable middle ground could be to make an SVG exporter that supports more than just pen-paths, but still has "known limitations" that throw warnings to the user. Such a library (likely an addon) would be unable to reproduce `clip()` in the near term, for example, but would handle visual stylings like fills, strokeWeights, and transparency, as [@mariuswatz](https://github.com/mariuswatz) requested [here](https://github.com/golanlevin/p5.plotSvg/issues/7).
> 
> More soon, Best Golan

I would be more than happy to collaborate on the issue.
So I thought about it and I feel that we should make a proper p5 addon and a SVG import.

How do you suggest we proceed? Ofcourse I will wait until you clean up the milestone you were talking about which is due tonight; should we collaborate on your library itself or should we build on top of yours?

Just to be sure enough, remaining things are:
- Fills + stroke weights in the exporter
- Text export
- SVG import
?

**Comment by NalinDalal** [comment]
I've been working on an SVG import addon that fills the `loadSVG()` gap. It's a standalone p5.js addon (not a core change) built on the `create-p5.js-addon` template.

**What it does**:
- `loadSVG(url)` — fetches an SVG file and parses it
- `parseSVG(svgString)` — parses SVG string into shape data
- `drawSVG(shapes)` — renders shapes using p5's `beginShape()`/`vertex()`/`endShape()` (works with both 2D and WEBGL)
- Shapes carry `id`, `className`, `style`, and `transform` for querying/modification

**Supported elements**: `<path>` (M/L/H/V/C/S/Q/T/A/Z), `<rect>` (with rounded corners), `<circle>`, `<ellipse>`, `<line>`, `<polyline>`, `<polygon>`, `<text>`, `<g>` with transforms

**Known limitations (not yet supported)**: gradients, filters, masks, clip-paths, `<use>/<defs>`, text-on-path, nested transforms — same scope as plotSvg's export

Repo: https://github.com/NalinDalal/p5.js-svg-export (see templates/svg-import/)

Rectangle seems to be working, would check for other shapes as well:

<img width="1440" height="900" alt="Image" src="https://github.com/user-attachments/assets/1238d9e5-e707-4034-b69e-9bcda74d7c52" />

<img width="1440" height="900" alt="Image" src="https://github.com/user-attachments/assets/27dbd014-4c91-49e7-a4dc-c72943258c71" />

**Comment by VANSH3104** [comment]
Hi everyone,

I'm also working on SVG support for p5.js as part of GSoC 2026. Over the past few months, I've been exploring a Shape-system-based approach for SVG export/import, and I've recently completed Version 1 of the export functionality.

Repo:
https://github.com/VANSH3104/p5.SVG.js/tree/flat

Example:
https://editor.p5js.org/vanshkabra05/sketches/45caLrfVy

Feedback thread:
https://discourse.processing.org/t/help-test-svg-export-for-p5-js-gsoc-2026/48850

A big thank you to everyone who's contributed ideas, libraries, references, and use cases in this thread over the years. Following the discussions here has been incredibly helpful while working on the project.

If you'd like to try it with one of your own sketches, I'd love to hear any thoughts, suggestions, or feedback. Thanks!
